### PR TITLE
Introduce new endpoint to allow fetching conversation metadata by qualified ids

### PR DESCRIPTION
**Pull request by akshaymankar**
The API will only support getting metadata of a given list of ids.
https://wearezeta.atlassian.net/wiki/spaces/CORE/pages/477660475/Federated+Conversations+Pagination

## Checklist

 - [x] The **PR Title** explains the impact of the change.
 - [x] The **PR description** provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] If end-points have been added or changed: the **endpoint / config-flag checklist** (see Wire-employee only backend [wiki page](https://github.com/zinfra/backend-wiki/wiki/Checklists)) has been followed.
 - [x] If a schema migration has been added, I ran **`make git-add-cassandra-schema`** to update the cassandra schema documentation.
 - [x] Section *Unreleased* of **CHANGELOG.md** contains the following bits of information:
   - [x] A line with the title and number of the PR in one or more suitable sub-sections.
   - [x] If /a: measures to be taken by instance operators.
   - [x] If /a: list of cassandra migrations.
   - [x] If public end-points have been changed or added: does nginz need upgrade?
   - [x] If internal end-points have been added or changed: which services have to be deployed in a specific order?

### CODE REVIEW

File: services/galley/src/Galley/API/Query.hs [hunk 1]
```diff
@@ -94,7 +100,7 @@ getConversation zusr cnv = do
   localDomain <- viewFederationDomain
   if qDomain cnv == localDomain
     then getUnqualifiedConversation zusr (qUnqualified cnv)
-    else getRemoteConversation zusr (toRemote cnv)
+    else getRemoteConversation zusr (toRemote cnv) -- TODO: This looks wrong, we shouldn't get a remote conversation is the requesting user is not part of that conversation.
```
Comment (akshaymankar) [inline services/galley/src/Galley/API/Query.hs:103]: I will create another PR for this.

Comment (akshaymankar) [reply services/galley/src/Galley/API/Query.hs]: #1713

**Review by jschaul — APPROVED** [review]

File: services/galley/src/Galley/API/Query.hs [hunk 2]
```diff
@@ -115,6 +121,30 @@ getRemoteConversations zusr remoteConvs = do
...
+  localDomain <- viewFederationDomain
+  let qualifiedZUser = Qualified zusr localDomain
+  let convsByDomain = partitionRemote remoteConvs
+  convs <- pooledForConcurrentlyN 8 convsByDomain $ \(remoteDomain, convIds) -> handleFailures remoteDomain convIds $ do
+    let req = FederatedGalley.GetConversationsRequest qualifiedZUser convIds
+        rpc = FederatedGalley.getConversations FederatedGalley.clientRoutes req
+    gcresConvs <$> executeFederated remoteDomain rpc
+  pure $ concatEithers convs
+  where
+    handleFailures :: Domain -> [ConvId] -> ExceptT FederationError Galley a -> Galley (Either [Qualified ConvId] a)
```
Comment (jschaul) [inline services/galley/src/Galley/API/Query.hs:135]: Is my assumption correct that this `handleFailures` function is not tested? Could we make this a top-level function and exercise a failure case in a test here? (Or maybe add a FUTUREWORK for this).

File: services/galley/src/Galley/Data.hs [hunk 3]
```diff
@@ -575,13 +577,45 @@ conversationIdRowsForPagination usr start (fromRange -> max) =
...
+remoteConversationIdOf usr = conncurrentPartitionM (pure . isJust <=< findRemoteConv)
+  where
+    findRemoteConv :: Remote ConvId -> m (Maybe (Identity UserId))
+    findRemoteConv remoteConvId = do
+      let (Qualified conv domain) = unTagged remoteConvId
+      query1 Cql.selectRemoteConvMembership (params Quorum (usr, domain, conv))
+
+-- | Like 'Control.Monad.Extra.partitionM', but works concurrently in 8 threads.
+conncurrentPartitionM :: forall m a. MonadUnliftIO m => (a -> m Bool) -> [a] -> m ([a], [a])
+conncurrentPartitionM predM xs = partitionEithers <$> pooledMapConcurrentlyN 8 bar xs
```
[the remark this answers is not in the document]
Comment (akshaymankar) [reply services/galley/src/Galley/Data.hs]: I decided that I didn't need that function if I could use the `IN` clause.

File: services/galley/src/Galley/Data.hs [hunk 4]
```diff
@@ -575,13 +577,45 @@ conversationIdRowsForPagination usr start (fromRange -> max) =
...
+-- given user and those which are not found.
+--
+-- Returns @(foundConvs, notFoundConvs)@.
+remoteConversationIdOf :: forall m. (MonadClient m, MonadLogger m, MonadUnliftIO m) => UserId -> [Remote ConvId] -> m ([Remote ConvId], [Remote ConvId])
+remoteConversationIdOf usr = conncurrentPartitionM (pure . isJust <=< findRemoteConv)
+  where
+    findRemoteConv :: Remote ConvId -> m (Maybe (Identity UserId))
+    findRemoteConv remoteConvId = do
+      let (Qualified conv domain) = unTagged remoteConvId
+      query1 Cql.selectRemoteConvMembership (params Quorum (usr, domain, conv))
```
[the remark this answers is not in the document]
Comment (akshaymankar) [reply services/galley/src/Galley/Data.hs]: This is much better, thanks!

File: services/galley/src/Galley/API/Query.hs (hunk 2 above)
Comment (akshaymankar) [reply services/galley/src/Galley/API/Query.hs:135]: It is exercised from integration tests when  Alice tries to get a remote conv on c.far-away.example.com, for which the federated call fails.

File: services/galley/src/Galley/Data.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (akshaymankar) [reply services/galley/src/Galley/Data.hs]: Wait, I misunderstood this and applied the change to both tables. I will revert it for the previous one.

File: services/galley/src/Galley/API/Query.hs (hunk 2 above)
Comment (jschaul) [reply services/galley/src/Galley/API/Query.hs:135]: ok cool, I had not seen that.

**Review by jschaul — APPROVED** [review]
This looks good now (with respect to cassandra queries), I think, at least in theory 😄

Another thought: Would it make sense to run one manual test (probably would take too long to run each time) which creates 1000 conversations in the remote table and then gets its metadata in one request? Just to ensure there are no CQL limits to an `IN [1..1000]` query?

**Comment by akshaymankar** [comment]
I wrote this test and it took more than a minute for galley to respond. Interestingly, registering a 1000 conversation, which took 1000 requests to galley, was very quick (~2s). I will check why the list request is slow.

```haskell
testBulkGet1000QualifiedConvs :: TestM ()
testBulkGet1000QualifiedConvs = do
  aliceQ <- randomQualifiedUser
  bobId <- randomId
  let remoteDomainA = Domain "a.far-away.example.com"
  let alice = qUnqualified aliceQ
      bobQ = Qualified bobId remoteDomainA

  remoteConvIds <- replicateM 1000 $ randomQualifiedId remoteDomainA


  let aliceAsOtherMember = OtherMember aliceQ Nothing roleNameWireAdmin
  preRegisterTime <- liftIO getCurrentTime
  print (preRegisterTime, "start registering" :: String)
  for_ remoteConvIds $ \conv ->
    registerRemoteConv conv bobQ Nothing (Set.fromList [aliceAsOtherMember])
  postRegisterTime <- liftIO getCurrentTime
  print (postRegisterTime, "finished registering" :: String)

  let aliceAsSelfMember = Member (qUnqualified aliceQ) Nothing False Nothing Nothing False Nothing False Nothing roleNameWireAdmin
      bobAsOtherMember = OtherMember bobQ Nothing roleNameWireAdmin
      mockConversations = map (\conv -> mkConv conv bobId aliceAsSelfMember [bobAsOtherMember]) remoteConvIds
      req = ListConversationsV2 . unsafeRange $ remoteConvIds
  opts <- view tsGConf
  (respAll, receivedRequests) <-
    withTempMockFederator'
      opts
      remoteDomainA
      ( \fedReq -> do
          let success = pure . F.OutwardResponseBody . LBS.toStrict . encode
          case F.domain fedReq of
            d | d == domainText remoteDomainA -> success $ GetConversationsResponse mockConversations
            _ -> assertFailure $ "Unrecognized domain: " <> show fedReq
      ) $ do
       preListTime <- liftIO getCurrentTime
       print (preListTime, "listing convs" :: String)
       listRes <- listConvsV2 alice req
       postListTime <- liftIO getCurrentTime
       print (postListTime, "retrieved the list" :: String)
       pure listRes

  convs <- responseJsonUnsafe <$> (pure respAll <!! const 200 === statusCode)

  liftIO $ do
    let expectedFound = sortOn cnvQualifiedId mockConversations
        actualFound = sortOn cnvQualifiedId $ crFound convs
    assertEqual "found conversations" expectedFound actualFound

    -- Assumes only one request is made
    let [requestedConvIds] = sort $ fmap (FederatedGalley.gcrConvIds . fromJust . decode @FederatedGalley.GetConversationsRequest . (LBS.fromStrict . F.body) . fromJust . F.request)  receivedRequests
    assertEqual "only locally found conversations should be queried" (sort $ qUnqualified <$> remoteConvIds) requestedConvIds

    let expectedNotFound = []
        actualNotFound = sort $ crNotFound convs
    assertEqual "not founds" expectedNotFound actualNotFound
    assertEqual "failures" [] (crFailed convs)
```

**Comment by akshaymankar** [comment]
With the power of lots of print statements, I could pin-point this to the part which grpcCall, so either our mock server is encoding the response too slowly or our grpc client is decoding it too slowly. Either way, this taking more than a minute is not acceptable, so I will keep digging. Another thing, I noted was this doesn't happen when we return one or two conversations.

**Pull request merged**